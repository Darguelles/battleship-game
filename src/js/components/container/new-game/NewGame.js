import React, {Component} from "react";
import {Button, Icon, Input, Modal, Row} from 'react-materialize';
import PropTypes from 'prop-types';
import {withRouter} from "react-router-dom";

class NewGame extends Component {

    constructor(props) {
        super(props);
        this.state = {
            gameName: '',
            players: '1',
            attempts: 'INFINITE',
            attemptsFieldBlocked: true,
        };
        this.handleInputChange = this.handleInputChange.bind(this);
        this.unlockAttemptFields = this.unlockAttemptFields.bind(this);
        this.startGame = this.startGame.bind(this);
    }

    handleInputChange(event) {
        const target = event.target;
        this.setState({[target.name]: target.value});
    }

    unlockAttemptFields(event) {
        let currentAttempts = event.target.value;
        let locked = true;
        if (currentAttempts == '0') {
            locked = false;
        }
        this.setState({
            attempts: currentAttempts,
            attemptsFieldBlocked: locked
        })
    }

    startGame() {
        let playerInfo = {
            'playerName': this.state.gameName,
            'attempts': this.state.attempts,
            'startTime': new Date(),
            'hits': 0,
            'failures': 0
        };
        if (playerInfo.playerName !== '' && playerInfo.attempts !== '') {
            window.localStorage.setItem('playerInfo', JSON.stringify(playerInfo));
            this.props.history.push("play");
        }
    }

    render() {
        return (
            <Modal header='Create game'
                   trigger={<Button waves='light' className={'button'}>New game<Icon left>play_arrow</Icon></Button>}
                   actions={<Button waves='light' className={'button'} onClick={this.startGame}>Play!<Icon
                       left>done</Icon></Button>}>

                <Row>
                    <Input id={'game_name'} s={12} l={6} m={12} label="Game name" value={this.state.gameName}
                           name={'gameName'}
                           onChange={this.handleInputChange} validate={true}/>
                    <Input id={'players'} s={12} l={6} m={12} defaultValue='1' type="number" label="Players"
                           name={'players'}
                           value={this.state.players} onChange={this.handleInputChange}/>
                    <Input id={'level-select'} s={12} l={6} m={12} type='select' label="Level" defaultValue='2'
                           value={this.state.attempts}
                           onChange={this.unlockAttemptFields}>
                        <option disabled value='Choose...'>Choose level</option>
                        <option value='INFINITE'>Easy</option>
                        <option value='100'>Medium</option>
                        <option value='50'>Hard</option>
                        <option value='0'>Custom</option>
                    </Input>
                    <Input id={'attempts'} s={12} l={6} m={12} label="Attempts" name={'attempts'}
                           disabled={this.state.attemptsFieldBlocked} value={this.state.attempts}
                           onChange={this.handleInputChange}/>
                </Row>

            </Modal>
        )
    }

}

NewGame.propTypes = {
    history: PropTypes.object,
};

export default withRouter(NewGame);