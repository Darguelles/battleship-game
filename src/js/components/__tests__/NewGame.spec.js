import {configure, shallow} from 'enzyme';
import React from 'react';
import NewGame from '../container/new-game/NewGame'
import renderer from 'react-test-renderer';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';


configure({ adapter: new Adapter() });

describe('NewGame > container behavior', () => {

    it('should renders correctly', () => {
        const rendered = renderer.create(
            <NewGame/>
        );

        expect(rendered.toJSON()).toMatchSnapshot();
    });

    it('should retrieve the correct value when some option is selected', () => {
        const component = shallow(<NewGame/>);
        component.find('#level-select').simulate('change', { target: {
            value: '50' }
        });

        expect(component.find('#level-select').props().value).toBe("50");
    });

    it('should pass a selected value to the onChange function', () => {
        const component = shallow(<NewGame/>);
        component.find('#level-select').simulate('change', { target: {
            value: '50' }
        });

        expect(toJson(component)).toMatchSnapshot();
    });

    it('should call unlockAttemptFields() on level select change', () => {
        const spy = jest.spyOn(NewGame.prototype, 'unlockAttemptFields');
        const wrapper = shallow(<NewGame/>);
        wrapper.find('#level-select').simulate('change', { target: {
            value: '50' }
        });
        expect(spy).toHaveBeenCalled();
    });

});

describe('NewGame > container logic', () => {

    it('should enable attempts input when Custom level is selected', () => {
        const component = shallow(<NewGame/>);

        component.find('#level-select').simulate('change', { target: {
            value: '0' }
        });

        const attempts = component.find('#attempts')
        expect(attempts.prop('disabled')).toEqual(false);
    });

    it('should disable attempts input when Custom level is not selected', () => {
        const component = shallow(<NewGame/>);
        const attempts = component.find('#attempts')

        component.find('#level-select').simulate('change', { target: {
            value: '50' }
        });

        expect(attempts.prop('disabled')).toEqual(true);
    })

});