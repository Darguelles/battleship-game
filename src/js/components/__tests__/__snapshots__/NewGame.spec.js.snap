// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewGame > container behavior should pass a selected value to the onChange function 1`] = `
<Modal
  actions={
    Array [
      <Button
        flat={true}
        modal="close"
        node="button"
        waves="light"
      >
        Close
      </Button>,
    ]
  }
  bottomSheet={false}
  fixedFooter={false}
  header="Create game"
  modalOptions={Object {}}
  trigger={
    <Button
      className="button"
      node="button"
      waves="light"
    >
      New game
      <Icon
        left={true}
      >
        save
      </Icon>
    </Button>
  }
>
  <Row>
    <Input
      checked={false}
      id="game_name"
      l={6}
      label="Game name"
      m={12}
      onChange={[Function]}
      s={12}
      type="text"
      value=""
    />
    <Input
      checked={false}
      defaultValue="1"
      id="players"
      l={6}
      label="Players"
      m={12}
      onChange={[Function]}
      s={12}
      type="number"
      value=""
    />
    <Input
      checked={false}
      defaultValue="2"
      id="level-select"
      l={6}
      label="Level"
      m={12}
      onChange={[Function]}
      s={12}
      type="select"
      value="50"
    >
      <option
        disabled={true}
        value="Choose..."
      >
        Choose level
      </option>
      <option
        value="INFINITE"
      >
        Easy
      </option>
      <option
        value="100"
      >
        Medium
      </option>
      <option
        value="50"
      >
        Hard
      </option>
      <option
        value="0"
      >
        Custom
      </option>
    </Input>
    <Input
      checked={false}
      disabled={true}
      id="attempts"
      l={6}
      label="Attempts"
      m={12}
      onChange={[Function]}
      s={12}
      type="text"
      value="50"
    />
  </Row>
</Modal>
`;

exports[`NewGame > container behavior should renders correctly 1`] = `
<button
  className="btn waves-effect waves-light button"
  disabled={false}
  onClick={[Function]}
>
  New game
  <i
    className="material-icons left"
  >
    save
  </i>
</button>
`;
